sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala

```diff
@@ -376,14 +376,10 @@ class HDFSFileCatalog(
         HadoopFsRelation.shouldFilterOut(name)
       }
 
-      val (dirs, files) = statuses.partition(_.isDirectory)
+      val (_, files) = statuses.partition(_.isDirectory)
 
       // It uses [[LinkedHashSet]] since the order of files can affect the results. (SPARK-11500)
-      if (dirs.isEmpty) {
-        mutable.LinkedHashSet(files: _*)
-      } else {
-        mutable.LinkedHashSet(files: _*) ++ listLeafFiles(dirs.map(_.getPath))
-      }
+      mutable.LinkedHashSet(files: _*)
```

Comment:
Also, I believe there is another method in `HadoopFsRelation` companion object to list up files parallely. This will use this method based on a threshold. I think that should be also corrected if it is really problematic.